File: src/common/py/core/messaging/meta/message_meta_information_list.py
import dataclasses
import threading
import time
import typing

from ..message import Trace
from .message_meta_information import MessageMetaInformation


class MessageMetaInformationList:
    """
    List to store message meta information objects.

    Notes:
        The list is thread-safe.
    """

    @dataclasses.dataclass(frozen=True)
    class _Entry:
        meta_information: MessageMetaInformation

        timeout: float = 0.0
        timestamp: float = dataclasses.field(default_factory=lambda: time.time())

        def has_timed_out(self) -> bool:
            """
            Whether the message has timed out.
            """
            return (
                time.time() - self.timestamp > self.timeout
                if self.timeout > 0.0
                else False
            )

    def __init__(self):
        self._list: typing.Dict[Trace, MessageMetaInformationList._Entry] = {}

        self._lock = threading.Lock()

    def add(self, unique: Trace, meta: MessageMetaInformation, timeout: float) -> None:
        """
        Adds a new entry to the list.

        Args:
            unique: The unique trace identifying the message.
            meta: The message meta information.
            timeout: A timeout (in seconds) after which a message is deemed timed out.
        """
        with self._lock:
            if unique not in self._list:
                self._list[unique] = MessageMetaInformationList._Entry(
                    meta, timeout=timeout
                )

    def remove(self, unique: Trace) -> None:
        """
        Removes an entry from the list.

        Args:
            unique: The unique trace identifying the message.
        """
        with self._lock:
            if unique in self._list:
                self._list.pop(unique)

    def find(self, unique: Trace) -> MessageMetaInformation | None:
        """
        Finds an entry associated with the given ``unique``.

        Args:
            unique: The unique trace identifying the message.

        Returns:
            The found meta information, if any.
        """
        entry = self._find(unique)
        return entry.meta_information if entry is not None else None

    def find_timed_out_entries(self) -> typing.List[Trace]:
        """
        Finds all entries that have timed out already.

        Returns:
            A list of all timed out entries.
        """
        with self._lock:
            return [
                unique for unique, entry in self._list.items() if entry.has_timed_out()
            ]

    def _find(self, unique: Trace) -> _Entry | None:
        with self._lock:
            if unique in self._list:
                return self._list[unique]

        return None

File: src/common/web/core/messaging/meta/MessageMetaInformationList.ts
import { MessageMetaInformation } from "./MessageMetaInformation";
import { type Trace } from "../Message";

class MessageMetaInformationListEntry {
    public constructor(readonly metaInformation: MessageMetaInformation,
                       readonly timeout: number = 0.0,
                       readonly timestamp: number = Date.now() / 1000.0) {
    }

    /**
     * Whether the message has timed out.
     */
    public hasTimedOut(): boolean {
        return this.timeout > 0.0 ? (Date.now() / 1000.0) - this.timestamp > this.timeout : false;
    }
}

interface MessageMetaInformationListType {
    [key: Trace]: MessageMetaInformationListEntry;
}

/**
 * List to store message meta information objects.
 */
export class MessageMetaInformationList {
    private readonly _list: MessageMetaInformationListType = {};

    /**
     * Adds a new entry to the list.
     *
     * @param unique - The unique trace identifying the message.
     * @param meta - The message meta information.
     * @param timeout - A timeout (in seconds) after which a message is deemed timed out.
     */
    public add(unique: Trace, meta: MessageMetaInformation, timeout: number): void {
        if (!(unique in this._list)) {
            this._list[unique] = new MessageMetaInformationListEntry(meta, timeout);
        }
    }

    /**
     * Removes an entry from the list.
     *
     * @param unique - The unique trace identifying the message.
     */
    public remove(unique: Trace): void {
        if (unique in this._list) {
            delete this._list[unique];
        }
    }

    /**
     * Finds an entry associated with the given ``unique``.
     *
     * @param unique - The unique trace identifying the message.
     *
     * @returns - The found meta information, if any.
     */
    public find(unique: Trace): MessageMetaInformation | null {
        if (unique in this._list) {
            return this._list[unique].metaInformation;
        }

        return null;
    }

    /**
     * Finds all entries that have timed out already.
     *
     * @returns - A list of all timed out entries.
     */
    public findTimedOutEntries(): Trace[] {
        let entries: Trace[] = [];
        for (const [unique, entry] of Object.entries(this._list)) {
            if (entry.hasTimedOut()) {
                entries.push(unique);
            }
        }
        return entries;
    }
}
